### PR TITLE
Add the capability to publish to InfluxDB

### PR DESCRIPTION
**Pull request by besanradwan**
This code aims to add the capability for PKB to publish directly to InfluxDB

The publishing of data from PKB benchmarks can done either by specifiying a pre-existing Influx DB publisher or with none specified it will create a new one or re-use a pre-existing DB called perfkit. By default if no uri is specified it will publish the data to localhost.

### CODE REVIEW

File: perfkitbenchmarker/publisher.py [hunk 1]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+                                                             'official',
+                                                             'owner',
+                                                             'run_uri',
+                                                             'sample_uri')})))
+      field_set = '%s_%s=%s' % (self.FormatSampleForInfluxDB({k: sample[k]
+                                for k in ('metric', 'unit', 'value')}))
+      samples_constructed_body = (' '.join((','.join((measurement,
+                                           tag_set))),
+                                           (' '.join((field_set,
+                                                     timestamp))))) + "\n"
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: Use local variables to decompose these 'joins' and the nested arguments

File: perfkitbenchmarker/publisher.py [hunk 2]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+
+  def PublishSamples(self, samples):
+    successful_http_request_codes = [200, 202, 204]
+    sample = list(samples)
+    number_of_samples = 0
+    for sample in samples:
+      number_of_samples += 1
+      timestamp = int((10 ** 9) * sample['timestamp'])
+      measurement = 'perfkitbenchmarker'
+      tag_set_metadata = \
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: nitpick: Prefer hanging indentation for style consistency of the codebase. See: https://www.python.org/dev/peps/pep-0008/#indentation

File: perfkitbenchmarker/publisher.py [hunk 3]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+      tag_set = ','.join((tag_set_metadata, self.
+                          FormatSampleForInfluxDB({k: sample[k]
+                                                   for k in ('test',
+                                                             'official',
+                                                             'owner',
+                                                             'run_uri',
+                                                             'sample_uri')})))
+      field_set = '%s_%s=%s' % (self.FormatSampleForInfluxDB({k: sample[k]
+                                for k in ('metric', 'unit', 'value')}))
+      samples_constructed_body = (' '.join((','.join((measurement,
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: Extract the logic to construct the request body for the InfluxDB to its own method.

File: perfkitbenchmarker/publisher.py [hunk 4]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+    conn.request('POST', '/query?' + body, headers)
+    response = conn.getresponse()
+    response_status = response.status
+    response_response = response.response
+    return response_status, response_response
+
+  def _WriteData(self, influx_uri, influx_db_name, data):
+    body = data
+    headers = {"Content-type": "application/octet-stream"}
+    conn = httplib.HTTPConnection(influx_uri)
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: Is it possible to re-use the connection instead of building a new one.

File: perfkitbenchmarker/publisher.py [hunk 5]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
 
+class InfluxDBPublisher(SamplePublisher):
+
+  def __init__(self, influx_uri=None, influx_db_name=None):
+    # set to default above in flags unless changed
+    self.influx_uri = influx_uri
+    self.influx_db_name = influx_db_name
+
+  def PublishSamples(self, samples):
+    successful_http_request_codes = [200, 202, 204]
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: HTTP error handling should probably happen within the method that makes the request.

File: perfkitbenchmarker/publisher.py (hunk 3 above)
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: Can you post a real example in the comments of how the final request body looks like? It's hard to follow what is the format of this string.

File: perfkitbenchmarker/publisher.py [hunk 6]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+    sample = list(samples)
+    number_of_samples = 0
+    for sample in samples:
+      number_of_samples += 1
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: No need to count the samples. Simply using `len(samples)` should return the number of samples.

File: perfkitbenchmarker/publisher.py [hunk 7]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+                    'Request could not be completed due to: ',
+                    create_db_response)
+
+    logging.debug('writing samples to publisher: writing',
+                  number_of_samples, ' samples.')
+    self._writeData(self.influx_uri, self.influx_db_name,
+                    samples_constructed_body)
+    logging.debug(create_db_status,
+                  'Request could not be completed due to: ',
+                  create_db_response)
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: Won't this always log that the request could not be completed?

File: perfkitbenchmarker/publisher.py [hunk 8]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+    self.influx_db_name = influx_db_name
+
+  def PublishSamples(self, samples):
+    successful_http_request_codes = [200, 202, 204]
+    sample = list(samples)
+    number_of_samples = 0
+    for sample in samples:
+      number_of_samples += 1
+      timestamp = int((10 ** 9) * sample['timestamp'])
+      measurement = 'perfkitbenchmarker'
```
Comment (vhe182) [inline perfkitbenchmarker/publisher.py]: string.join(measurement) doesn't destroy the measurement string, why not move the initialization of measurement before the for loop?

File: perfkitbenchmarker/publisher.py [hunk 9]
```diff
@@ -640,6 +650,81 @@ def _deDotKeys(self, res):
...
+                                                             'sample_uri')})))
+      field_set = '%s_%s=%s' % (self.FormatSampleForInfluxDB({k: sample[k]
+                                for k in ('metric', 'unit', 'value')}))
+      samples_constructed_body = (' '.join((','.join((measurement,
+                                           tag_set))),
+                                           (' '.join((field_set,
+                                                     timestamp))))) + "\n"
+
+    create_db_status, create_db_response = self. \
+        _createInfluxDB(self.influx_uri, self.influx_db_name)
```
Comment (vhe182) [inline perfkitbenchmarker/publisher.py]: _createInfluxDB() isn't a function in this file.  Did you mean _createDB() ?

**Comment by besanradwan** [comment]
@meteorfox @vhe182

File: perfkitbenchmarker/publisher.py [hunk 10]
```diff
@@ -106,6 +108,14 @@
     'of all samples as metadata. Multiple key-value pairs may be specified '
     'by separating each pair by commas.')
 
+flags.DEFINE_string(
+    'influx_uri', 'localhost:8086',
+    'The Influx DB endpoint for queries address and port')
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: nitpick: The help message readability could be improved, I think you can remove 'endpoint for queries'.

Also, I think it needs clarification on what's the format of the string it expects. It's not clear if passing 'http://hostname:port/' is valid, or if only 'hostname:port' is expected.

File: perfkitbenchmarker/publisher.py [hunk 11]
```diff
@@ -640,6 +650,90 @@ def _deDotKeys(self, res):
...
+    params = data
+    header = {"Content-type": "application/octet-stream"}
+    conn = httplib.HTTPConnection(self.influx_uri)
+    conn.request('POST', '/write?' + 'db=' + self.influx_db_name, params,
+                 headers=header)
+    response = conn.getresponse()
+    response_status = response.status
+    response_response = response.response
+    if response_status in successful_http_request_codes:
+      logging.debug('writing samples to publisher: writing samples.')
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: nitpick: Log message should start with capital letter for consistency.

File: perfkitbenchmarker/publisher.py [hunk 12]
```diff
@@ -640,6 +650,90 @@ def _deDotKeys(self, res):
...
+    else:
+      logging.debug(response_status,
+                    'Request could not be completed due to: ',
+                    response_response)
+      raise httplib.HTTPException
+
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: I'm not 100% sure if Python cleans up the connection, but I think it wouldn't hurt to close the connection after we are done using it.

File: perfkitbenchmarker/publisher.py [hunk 13]
```diff
@@ -640,6 +650,90 @@ def _deDotKeys(self, res):
...
+    conn = httplib.HTTPConnection(self.influx_uri)
+    conn.request('POST', '/write?' + 'db=' + self.influx_db_name, params,
+                 headers=header)
+    response = conn.getresponse()
+    response_status = response.status
+    response_response = response.response
+    if response_status in successful_http_request_codes:
+      logging.debug('writing samples to publisher: writing samples.')
+    else:
+      logging.debug(response_status,
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: nitpick: Instead of using local variables `response_status` and `response_response`, simply use the response object directly (i.e. `response.status`, `response.response`)

File: perfkitbenchmarker/publisher.py [hunk 14]
```diff
@@ -640,6 +650,90 @@ def _deDotKeys(self, res):
...
+  def _CreateDB(self):
+    successful_http_request_codes = [200, 202, 204]
+    header = {'Content-type': 'application/x-www-form-urlencoded',
+              'Accept': 'text/plain'}
+    params = urllib.urlencode({'q': 'CREATE DATABASE ' + self.influx_db_name})
+    conn = httplib.HTTPConnection(self.influx_uri)
+    conn.request('POST', '/query?' + params, headers=header)
+    response = conn.getresponse()
+    response_status = response.status
+    response_response = response.response
```
Comment (meteorfox) [inline perfkitbenchmarker/publisher.py]: See command below about `response_status` and `response_response`

**Comment by besanradwan** [comment]
@meteorfox @vhe182 please review.

File: tests/publisher_test.py [hunk 15]
```diff
@@ -404,3 +404,94 @@ def testUsesUnionOfMetaKeys(self):
...
+                      'run_uri': 'bba3', 'sample_uri': 'bb', 'timestamp': 55,
+                      'metadata': {}},
+                     {'test': 'testa', 'metric': '3', 'official': 47.0,
+                     'value': 'non', 'unit': 'us', 'owner': 'Rackspace',
+                      'run_uri': '5rtw', 'sample_uri': '5r', 'timestamp': 123}]
+      formated_samples = []
+      for data in sample_data:
+        formated_samples.append(self.mock_db._ConstructSample(sample))
+        self.assertEqual(formated_samples, 'perfkitbenchmarker,metric=1,\
+                                          official=1.0,owner=Rackspace,\
```
Comment (meteorfox) [inline tests/publisher_test.py]: Either use multi-line string wrapped with `( )`, instead of breaking the line with `\`, or perhaps consider using a data file for this string since it's very long. See the `tests/linux_benchmarks/unixbench_benchmark_test.py` for an example. https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/tests/linux_benchmarks/unixbench_benchmark_test.py#L29-L33

**Comment by besanradwan** [comment]
@meteorfox @gareth-ferneyhough please review

File: tests/publisher_test.py [hunk 16]
```diff
@@ -404,3 +404,90 @@ def testUsesUnionOfMetaKeys(self):
     rows = list(reader)
     self.assertEqual(['key1', 'key3'], reader.fieldnames[-2:])
     self.assertEqual(3, len(rows))
+
+
+  class InfluxDBPublisherTestCase(unittest.TestCase):
+    def setUp(self):
+      self.db_name = 'test_db'
+      self.db_uri = 'test'
+      self.mock_db = publisher.InfluxDBPublisher(self.db_uri, self.db_name)
```
Comment (meteorfox) [inline tests/publisher_test.py]: Since this is the actual instance of a publisher, I think `mock` shouldn't be included in the variable name, since it can be confused for a mock object.

File: tests/publisher_test.py [hunk 17]
```diff
@@ -404,3 +404,90 @@ def testUsesUnionOfMetaKeys(self):
...
+                         '123000000000\nperfkitbenchmarker,metric=2,'
+                         'official=14.0,owner=Rackspace,run_uri=bba3,'
+                         'test=testb,unit=MB,sample_uri=bb value=non '
+                         '55000000000\nperfkitbenchmarker,metric=3,'
+                         'official=47.0,owner=Rackspace,run_uri=5rtw,'
+                         'test=testa,unit=us,sample_uri=5r '
+                         'value=non 123000000000')
+      for data in sample_data:
+        formated_samples.append(self.mock_db._ConstructSample(sample))
+        self.assertEqual(formated_samples, expected_output)
```
Comment (meteorfox) [inline tests/publisher_test.py]: I'm not sure I understand how this test works.

Wouldn't this be calling an assertEqual between the `formatted_samples` list and the string `expected_output`? Also, the list `formatted_samples` changes in every iteration as the samples are added but the string `expected output` remains constant. This test would fail every time.

File: tests/publisher_test.py [hunk 18]
```diff
@@ -404,3 +404,90 @@ def testUsesUnionOfMetaKeys(self):
     rows = list(reader)
     self.assertEqual(['key1', 'key3'], reader.fieldnames[-2:])
     self.assertEqual(3, len(rows))
+
+
+  class InfluxDBPublisherTestCase(unittest.TestCase):
```
Comment (meteorfox) [inline tests/publisher_test.py]: This is the line why your tests aren't running.

This whole class is nested inside the CSVPublisherTestCase class. Please unindent and recommit. Thanks!

**Comment by besanradwan** [comment]
@meteorfox @vhe182 please review

File: tests/publisher_test.py [hunk 19]
```diff
@@ -404,3 +404,112 @@ def testUsesUnionOfMetaKeys(self):
...
+                    'run_uri': '5rtw', 'sample_uri': '5r', 'timestamp': 123}]
+    formated_samples = []
+    s1 = ('perfkitbenchmarker,sample_uri=33,owner=Rackspace,official=1.0,'
+          'run_uri=323,metric=1,test=testc,unit=MB,bar=foo,more_info=2,'
+          'info=1 value=non 123000000000 \n perfkitbenchmarker,'
+          'sample_uri=bb,owner=Rackspace,official=14.0,run_uri=bba3,metric=2,'
+          'test=testb,unit=MB value=non 55000000000 \n perfkitbenchmarker,'
+          'sample_uri=5r,owner=Rackspace,official=47.0,run_uri=5rtw,metric=3,'
+          'test=testa,unit=us value=non 123000000000')
+    expected_sample_format = self.sortString(s1)
```
Comment (meteorfox) [inline tests/publisher_test.py]: I don't think you should sort the expected string. It needs to match the line protocol ordered as defined in Influx here:
https://docs.influxdata.com/influxdb/v1.2/write_protocols/line_protocol_reference/

Also, based on that reference, there shouldn't be whitespace between the newline characters `\n`.

File: tests/publisher_test.py [hunk 20]
```diff
@@ -404,3 +404,112 @@ def testUsesUnionOfMetaKeys(self):
...
+    expected_2 = expected_1.split()
+    expected_3 = ','.join(expected_2)
+    expected_4 = sorted(expected_3.split(','))
+    return expected_4
+
+  def sortList(self, stuff):
+    new_string = []
+    for i in stuff:
+      if "metadata" in i:
+        i = i.replace('metadata={', '').replace('}',
```
Comment (meteorfox) [inline tests/publisher_test.py]: Isn't this removing `metadata`?

File: tests/publisher_test.py [hunk 21]
```diff
@@ -404,3 +404,112 @@ def testUsesUnionOfMetaKeys(self):
...
+          'sample_uri=5r']
+    s2 = ['owner=Rackspace', 'unit=MB', 'run_uri=bba3', 'test=testb',
+          'timestamp=55', 'metric=2', 'official=14.0', 'metadata={}',
+          'value=non', 'sample_uri=bb']
+    s3 = ['owner=Rackspace', 'unit=MB', 'run_uri=323', 'test=testc',
+          'timestamp=123', 'metric=1', 'official=1.0',
+          "metadata={'info': '1','bar': 'foo', 'more_info': '2'}", 'value=non',
+          'sample_uri=33']
+
+    self.assertEqual(self.sortList(sorted(no_meta_keys)),
```
Comment (meteorfox) [inline tests/publisher_test.py]: Maybe use `assertDictEqual(...)` instead? 
Ref: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertDictEqual

File: tests/publisher_test.py [hunk 22]
```diff
@@ -404,3 +404,112 @@ def testUsesUnionOfMetaKeys(self):
...
+    s1 = ('perfkitbenchmarker,sample_uri=33,owner=Rackspace,official=1.0,'
+          'run_uri=323,metric=1,test=testc,unit=MB,bar=foo,more_info=2,'
+          'info=1 value=non 123000000000 \n perfkitbenchmarker,'
+          'sample_uri=bb,owner=Rackspace,official=14.0,run_uri=bba3,metric=2,'
+          'test=testb,unit=MB value=non 55000000000 \n perfkitbenchmarker,'
+          'sample_uri=5r,owner=Rackspace,official=47.0,run_uri=5rtw,metric=3,'
+          'test=testa,unit=us value=non 123000000000')
+    expected_sample_format = self.sortString(s1)
+    for data in sample_data:
+      formated_samples.append(self.test_db._ConstructSample(data))
```
Comment (meteorfox) [inline tests/publisher_test.py]: Since `_ConstructSample(data)` returns a string, why not simply test for that? No need to join them or append them in a list.

If you want to test the expected data format that gets sent to Influx, I'd suggest then to stub out `_CreateDB()` and `_WriteData()` and add a test for `_Publish()` to assert that `_WriteData` gets called with the data format you expect.

**Comment by besanradwan** [comment]
@meteorfox @vhe182 Please review

**Comment by besanradwan** [comment]
@gareth-ferneyhough please review

**Comment by meteorfox** [comment]
@gareth-ferneyhough PTAL, I intend to merge this PR to master. It adds a publisher implementation for InfluxDB. Thanks!

File: tests/publisher_test.py [hunk 23]
```diff
@@ -404,3 +404,143 @@ def testUsesUnionOfMetaKeys(self):
...
+         'run_uri=bba3,sample_uri=bb,metric=2,unit=MB value=non 55000000000'),
+        ('perfkitbenchmarker,test=testa,official=47.0,owner=Rackspace,'
+         'run_uri=5rtw,sample_uri=5r,metric=3,unit=us value=non 123000000000')
+    ]
+
+    mock_publish_method.return_value = None
+    self.test_db.PublishSamples(samples)
+    mock_publish_method.assert_called_once_with(expected)
+
+
```
Comment (gareth-ferneyhough) [inline tests/publisher_test.py]: Remove extra blank line.

File: perfkitbenchmarker/publisher.py [hunk 24]
```diff
@@ -645,6 +657,92 @@ def _deDotKeys(self, res):
...
+    for k, v in sample.iteritems():
+      if v == '':
+        v = '\\"\\"'
+      v = str(v)
+      v = v.replace(',', '\,')
+      v = v.replace(' ', '\ ')
+      key_value_pairs.append('%s=%s' % (k, v))
+    return key_value_pairs
+
+  def _CreateDB(self):
```
Comment (gareth-ferneyhough) [inline perfkitbenchmarker/publisher.py]: It looks like this is always called? What if you wish to publish to an existing database?

Comment (besanradwan) [inline perfkitbenchmarker/publisher.py]: Hey @gareth-ferneyhough I made the requested change to remove the extra blank line, however as for why _CreateDB() is always called: this is because it is idempotent in the sense that if the DB already exists it will simply return a 200 code without re-creating it.

Comment (gareth-ferneyhough) [inline perfkitbenchmarker/publisher.py]: Great. Can you just add a comment stating this fact? Then everything should be good to go.

Comment (besanradwan) [inline perfkitbenchmarker/publisher.py]: Done! @gareth-ferneyhough thanks for the feedback.

**Comment by gareth-ferneyhough** [comment]
All good. Thank you again.
